Include/Beman/InplaceVector/inplace_vector.hpp

```diff
@@ -0,0 +1,755 @@
+  }
+  constexpr const_reverse_iterator rend() const noexcept {
+    return const_reverse_iterator{this->begin()};
+  }
+  constexpr const_iterator cbegin() const noexcept { return this->begin(); }
+  constexpr const_iterator cend() const noexcept { return this->end(); }
+  constexpr const_reverse_iterator crbegin() const noexcept {
+    return const_reverse_iterator{this->end()};
+  }
+
+  constexpr const_reverse_iterator crend() const noexcept {
+    return const_reverse_iterator{this->begin()};
+  };
+
+  // [containers.sequences.inplace.vector.members] size/capacity
+  constexpr bool empty() const noexcept { return size() == 0; };
+  constexpr size_type size() const noexcept { return this->size_; };
+  static constexpr size_type max_size() noexcept { return Capacity; };
+  static constexpr size_type capacity() noexcept { return Capacity; };
+  constexpr void resize(size_type sz) {
+    const auto diff = static_cast<std::ptrdiff_t>(sz - this->size());
+    if (diff < 0) {
+      std::destroy(this->begin() + sz, this->end());
+    } else {
+      if (Capacity < size) {
+        throw std::bad_alloc();
+      }
+      const iterator end = this->end();
+      base::uninitialized_value_construct(end, end + diff);
+    }
+    this->change_size(diff);
+  }; // freestanding-deleted
+  constexpr void resize(size_type sz, const T &c) {
+    const auto diff = static_cast<std::ptrdiff_t>(sz - this->size());
+    if (diff < 0) {
+      std::destroy(this->begin() + sz, this->end());
+    } else {
+      if (Capacity < sz) {
+        throw std::bad_alloc();
+      }
+      const iterator end = this->end();
+      std::uninitialized_fill(end, end + diff, c);
+    }
+    this->change_size(diff);
+  }; // freestanding-deleted
+  static constexpr void reserve(size_type sz) {
+    if (Capacity < sz) {
+      throw std::bad_alloc();
+    }
+  }; // freestanding-deleted
+  static constexpr void shrink_to_fit() noexcept {}
+
+  // [containers.sequences.inplace.vector.modifiers], modifiers
+  template <class... Args> constexpr reference emplace_back(Args &&...args) {
+    if (this->size() == Capacity) {
+      throw std::bad_alloc();
+    }
+    return this->unchecked_emplace_back(std::forward<Args>(args)...);
+  }; // freestanding-deleted
+
+  constexpr reference push_back(const T &x) {
+    if (this->size() == Capacity) {
+      throw std::bad_alloc();
+    }
+    return this->unchecked_emplace_back(x);
+  }; // freestanding-deleted
+  constexpr reference push_back(T &&x) {
+    if (this->size() == Capacity) {
+      throw std::bad_alloc();
+    }
+    return this->unchecked_emplace_back(std::move(x));
+  }; // freestanding-deleted
+  template <typename R>
+    requires container_compatible_range<T>
+  constexpr void append_range(R &&rg) {
+    auto first = rg.begin();
+    auto last = rg.end();
+    for (; first != last; ++first) {
+      emplace_back(*first);
+    }
+  }; // freestanding-deleted
+  constexpr void pop_back() {
```

Comment:
I'm also wondering about conditional noexcept here -- it's not noexcept in any form, in the standard text, but certainly could be if T is no throw destructible. Also it's not marked as freestanding deleted -- which I suppose is because no T on freestanding is allowed to throw? Worrying we might have some spec bugs here.

Reply:
Re: freestanding
The design in the standard is to avoid originating exceptions.  If a user type throws first, then that isn't an issue.  This comes to the forefront in std::visit, as it can throw if a variant is in the valueless state.  You can only get on the valueless state when an exception is thrown though.

As for the noexceptness, I think it isn't currently noexcept because there's a precondition on the container being non-empty.  I thought we generally assumed stuff was noexcept destructible.  I think this could be marked unconditionally noexcept.